dissect/util/stream.py

```diff
@@ -415,3 +424,115 @@ def _read(self, offset: int, length: int) -> bytes:
             run_idx += 1
 
         return b"".join(r)
+
+
+class OverlayStream(AlignedStream):
+    """Create a stream from another file-like object with the ability to overlay other streams or bytes.
+
+    Useful for patching large file-like objects without having to cache the entire contents.
+    First wrap the original stream in this class, and then call ``add()`` with the offset and data to overlay.
+
+    Args:
+        fh: The source file-like object.
+        size: The size the stream should be.
+        align: The alignment size.
+    """
+
+    def __init__(self, fh: BinaryIO, size: Optional[int] = None, align: int = STREAM_BUFFER_SIZE):
+        super().__init__(size, align)
+        self._fh = fh
+        self.overlays: dict[int, tuple[int, BinaryIO]] = {}
+        self._lookup: list[int] = []
+
+    def add(self, offset: int, data: Union[bytes, BinaryIO], size: Optional[int] = None) -> None:
+        """Add an overlay at the given offset.
+
+        Args:
+            offset: The offset in bytes to add an overlay at.
+            data: The bytes or file-like object to overlay
+            size: Optional size specification of the overlay, if it can't be inferred.
+        """
+        if not hasattr(data, "read"):
```

Comment:
I think this behaviour might be confusing. I would suggest that size overrides the size always and if no size has been given and the size cannot be inferred this would simply result in an error. Now the caller has to know whether the class will be able to 'infer' the size.

Reply:
What do you suggest?

Reply:
My suggestion:

- Attempt to infer the size using len() or seek(0,2)/tell() of size is None
- Override with specified size if any

This way you can also check if the size argument, if provided is invalid, in which case you can raise an exception.
I would not rely on an implicit "size" property, or at least provide an additional method/param type with the required interface type for that.